crates/hir_ty/src/lower.rs

```diff
@@ -845,13 +873,42 @@ impl<'a> TyLoweringContext<'a> {
             })
     }
 
-    fn lower_impl_trait(&self, bounds: &[Interned<TypeBound>]) -> ReturnTypeImplTrait {
+    fn lower_impl_trait(
+        &self,
+        bounds: &[Interned<TypeBound>],
+        func: FunctionId,
+    ) -> ReturnTypeImplTrait {
         cov_mark::hit!(lower_rpit);
         let self_ty =
             TyKind::BoundVar(BoundVar::new(DebruijnIndex::INNERMOST, 0)).intern(&Interner);
+        // XXX(iDawer): Can shifting mess with unsized_types? For now I better reinsure.
+        let outer_unsized_types = self.unsized_types.replace(Default::default());
```

Comment:
is this necessary?

Reply:
I'll recheck, test and leave a comment if it still needs.

Reply:
If I remember correctly I did mess up with Debrujin indices and tried random things (with this) to fix it. Seems not actual anymore.